modules/ting_search/modules/subsearch_kpi/subsearch_kpi.module

```diff
@@ -0,0 +1,220 @@
+
+  if (variable_get('do_you_mean_guess', FALSE) != 'kpi' || !$keys) {
+    return $messages;
+  }
+
+  // When a zero-hit search is executed and we're able to match "correct"
+  // keys, a new search is done with the "correct" keys.
+  // Then use the query-parameter original-keys to display what was done to
+  // the visitor.
+  $parameters = drupal_get_query_parameters();
+  if (isset($parameters['original-search'])) {
+    $msg = t('Did you mean "!keys?" which gave !num-results. The search was executed. The search for "!original-keys" returned 0 hits.', array(
+      '!original-keys' => $parameters['original-search'],
+      '!keys' => $keys,
+      '!num-results' => $results->numTotalObjects,
+    ));
+
+    $messages[] = '<div class="messages warning">' . $msg . '</div>';
+  }
+
+  if ($results->numTotalObjects == 0) {
+    // The search result returned zero-hits so we need to see if "correct" keys
+    // can yield a better result.
+    $search_result = drupal_static('ting_search_results');
+    // Find suggestions for "correct" search phrase.
+    $suggestion = subsearch_kpi_suggest_keys($keys);
+
+    if (!empty($suggestion)) {
+      // Do secondary search for "correct" keys.
+      $secondary_result = ting_search_do_secondary_search($suggestion);
+      // Only redirect to "correct" keys search if it has a positive result.
+      if ($secondary_result->numTotalObjects > 0) {
+        $params['query'] = array(
+          'original-search' => $search_result->search_key
+        );
+        drupal_goto('search/ting/' . strtolower($suggestion), $params);
+      }
+    }
+  }
+
+  if ($results->numTotalObjects <= variable_get('do_you_mean_case_4_minimum', 5)) {
+    $suggestion = subsearch_kpi_suggest_keys($keys);
+    if ($suggestion) {
+      // Do aditional search.
+      $suggested_result = ting_search_do_secondary_search($suggestion);
+      if ($suggested_result->numTotalObjects >= 10) {
+
+        // Set message!
+        $msg = t('The search for "!keys" returned !num-results hits. Did you mean "<a href="/search/ting/!suggested-keys">!suggested-keys</a>?" which will return !suggested-num-results hits.', array(
+          '!suggested-keys' => strtolower($suggestion),
+          '!suggested-num-results' => $suggested_result->numTotalObjects,
+          '!keys' => $keys,
+          '!num-results' => $results->numTotalObjects,
+        ));
+        $messages[] = '<div class="messages warning">' . $msg . '</div>';
+      }
+    }
+  }
+  return $messages;
+}
+
+/**
+ * Find the first suggestion for given keys.
+ *
+ * @param string $keys
+ *   The original keys that you want to find suggestion for.
+ *
+ * @return string
+ *   The suggestion.
+ */
+function subsearch_kpi_suggest_keys($keys) {
+
+  $kpi_index = db_select('kpi_index_popular_keys', 'k')
+    ->fields('k')
+    ->condition('numtotalsearches', variable_get('do_you_mean_minimum_searches', 10),'>')
+    ->orderBy('numtotalsearches', 'DESC')
+    ->execute();
+
```

Comment:
This query needs to take into accout the original key, so we do not return that same key again. This avoids suggesting the original key to the user again.